base/compiler/optimize.jl

```diff
@@ -570,9 +570,14 @@ function ipo_dataflow_analysis!(interp::AbstractInterpreter, ir::IRCode, result:
         return true
     end
 
-    function scan_non_dataflow_flags!(inst::Instruction)
+    function scan_non_dataflow_flags!(inst::Instruction, @nospecialize(stmt=inst[:stmt]))
         flag = inst[:flag]
-        all_effect_free &= (flag & IR_FLAG_EFFECT_FREE) != 0
+        if !isterminator(stmt)
```

Comment:
Let's add `&& stmt !== nothing` so that statement deletion is automatically considered effect free. Then we don't need this in the other place.

Reply:
I thought https://github.com/JuliaLang/julia/pull/51185/commits/b0f30b10aac134402fb7d9896562911bd2eac9fd is a more general solution because we may observe the IR flag in other places too. But I think only this postopt analysis and the statement deletion observes it, and since the latter already has a special case for `nothing`, we may want to have it here too.
```suggestion
        if !isterminator(stmt) && stmt !== nothing
```